diff --git a/.DS_Store b/.DS_Store
index 664c68d..213291a 100644
Binary files a/.DS_Store and b/.DS_Store differ
diff --git a/asm_dir/controller/main.c b/asm_dir/controller/main.c
index 9d11f77..b0785ee 100644
--- a/asm_dir/controller/main.c
+++ b/asm_dir/controller/main.c
@@ -75,22 +75,85 @@ int		record_label(int *i, char *src, t_asm_str **asm_str)
 	return (fl);
 }
 
+int 	record_arg_type(char arg_char, t_asm_str **asm_str, int a)
+{
+	if (arg_char == REG_CHAR)
+	{
+		(*asm_str)->type_arg[a] = REG_CODE;
+		(*asm_str)->size += REG_SIZE;
+		return (1);
+	}
+	if (arg_char == DIRECT_CHAR)
+	{
+		(*asm_str)->type_arg[a] = DIR_CODE;
+		(*asm_str)->size += DIR_SIZE((*asm_str)->label_size);
+		return (1);
+	}
+	(*asm_str)->type_arg[a] = IND_CODE;
+	(*asm_str)->size += IND_SIZE;
+	return (0);
+}
+
+int		parse_args(char *src, t_asm_str **asm_str, int len, int *temp)
+{
+	int 	n;
+	int 	a;
+	char	**args;
+
+	n = 0;
+	a = 0;
+	args = ft_strsplit(src, SEPARATOR_CHAR);
+	while (args[n])
+	{
+		*temp = 1;
+		len = 0;
+		while (args[n][len] && (args[n][len] == ' ' || args[n][len] == '\t'))
+			len++;
+		len += record_arg_type(args[n][len], asm_str, a);
+		if (ft_strlen_space(args[n]) == 1)
+			error_type(LEXIC, (*asm_str)->nb);
+		(*asm_str)->arg_cont[a] = ft_atoi(&args[n][len]);
+		if (args[n][len] == LABEL_CHAR)
+			(*asm_str)->pointer[a] = ft_strdup(&args[n][++len]);
+		a++;
+		n++;
+	}
+	free_2d_array(&args);
+	return (n);
+}
+
+int 	identify_operation(int len, char *src, t_asm_str **asm_str)
+{
+	int 	oper;
+	char	*op;
+
+	oper = 0;
+	op = ft_strnew(len);
+	ft_strncpy(op, src, len);
+	while (oper < 17)
+	{
+		if (ft_strcmp(g_op_tab[oper].name, op) == EQUAL)
+		{
+			(*asm_str)->op_code = g_op_tab[oper].op_code;
+			(*asm_str)->size = 1;
+			break ;
+		}
+		oper++;
+	}
+	ft_strdel(&op);
+	return (oper);
+}
+
 void	record_asm_str(char *src, t_asm_str **asm_str, int nb)
 {
 	int		i;
-	int		n;
-	int		a;
 	int		len;
-	char	*op;
-	char	**args;
 	int		fl;
 	int		temp;
 	int		oper;
 
 	temp = 0;
-	oper = 0;
 	i = 0;
-	a = 0;
 	(*asm_str)->nb = nb;
 	if (src[i] == LABEL_CHAR)
 		error_type(INC_LABELNAME, (*asm_str)->nb);
@@ -100,66 +163,20 @@ void	record_asm_str(char *src, t_asm_str **asm_str, int nb)
 	while (src[i] == ' ' || src[i] == '\t')
 		i++;
 	len = ft_wordlen(&src[i]);
-	op = ft_strnew(len);
-	ft_strncpy(op, &src[i], len);
-	while (oper < 17)
-	{
-		if (ft_strcmp(g_op_tab[oper].name, op) == EQUAL)
-		{
-			(*asm_str)->op_code = g_op_tab[oper].op_code;
-			(*asm_str)->size = 1;
-			break ;
-		}
-		oper++;
-	}
+	oper = identify_operation(len, &src[i], asm_str);
 	if (g_op_tab[oper].codage)
 	{
 		(*asm_str)->codage = 1;
 		(*asm_str)->size += 1;
 	}
 	(*asm_str)->label_size = g_op_tab[oper].label_size;
-	n = 0;
 	i += len + 1;
-	args = ft_strsplit(&src[i], SEPARATOR_CHAR);
-	while (args[n])
-	{
-		temp = 1;
-		len = 0;
-		while (args[n][len] && (args[n][len] == ' ' || args[n][len] == '\t'))
-			len++;
-		if (args[n][len] == REG_CHAR)
-		{
-			(*asm_str)->type_arg[a] = REG_CODE;
-			(*asm_str)->size += REG_SIZE;
-			len++;
-		}
-		else if (args[n][len] == DIRECT_CHAR)
-		{
-			(*asm_str)->type_arg[a] = DIR_CODE;
-			(*asm_str)->size += DIR_SIZE((*asm_str)->label_size);
-			len++;
-		}
-		else
-		{
-			(*asm_str)->type_arg[a] = IND_CODE;
-			(*asm_str)->size += IND_SIZE;
-		}
-		if (ft_strlen_space(args[n]) == 1)
-			error_type(LEXIC, (*asm_str)->nb);
-		(*asm_str)->arg_cont[a] = ft_atoi(&args[n][len]);
-		if (args[n][len] == LABEL_CHAR)
-		{
-			(*asm_str)->pointer[a] = ft_strdup(&args[n][++len]);
-		}
-		a++;
-		n++;
-	}
+	if (g_op_tab[oper].arg_nb != parse_args(&src[i], asm_str, len, &temp))
+		error_type(INC_ARGS, (*asm_str)->nb);
 	if (oper == 17)
 		error_type(OP_EXIST, (*asm_str)->nb);
 	if (!temp)
 		error_type(INC_ARGS, (*asm_str)->nb);
-	if (g_op_tab[oper].arg_nb != n)
-		error_type(INC_ARGS, (*asm_str)->nb);
 }
 
 char	is_empty_str(char *str)
@@ -243,13 +260,9 @@ void	write_cor_file_instructions(int fd_cor, t_asm_str *asm_str)
 			codage = calculate_codage(asm_str);
 			write(fd_cor, &codage, 1);
 		}
-		t_rotate.num = 0;
-		t_rotate.num = asm_str->arg_cont[i];
 		i = 0;
 		while ((i < 3) && asm_str->type_arg[i])
 		{
-			t_rotate.num = 0;
-			t_rotate.num = asm_str->arg_cont[i];
 			if (asm_str->type_arg[i] == REG_CODE)
 			{
 				write(fd_cor, &(asm_str->arg_cont[i]), REG_SIZE);
diff --git a/asm_dir/includes/asm.h b/asm_dir/includes/asm.h
index 7603f93..a526b5b 100644
--- a/asm_dir/includes/asm.h
+++ b/asm_dir/includes/asm.h
@@ -66,13 +66,6 @@ struct						s_asm_str
 	t_asm_str				*prev;
 };
 
-union						u_rotate
-{
-	int						num;
-	char					buf2[2];
-	char					buf4[4];
-}							t_rotate;
-
 /*
 ******************************** MAIN FUNCTIONS ********************************
 */
@@ -91,6 +84,7 @@ int							ft_wordlen(char *str);
 char						*cut_content(char *str);
 char						is_empty_str(char *str);
 void						error_type(char *line, int nb);
+void						free_2d_array(char ***arr);
 
 /*
 ******************************** UTIL FUNCTIONS ********************************
diff --git a/asm_dir/includes/corewar.h b/asm_dir/includes/corewar.h
index 4422be2..ffd4583 100644
--- a/asm_dir/includes/corewar.h
+++ b/asm_dir/includes/corewar.h
@@ -17,8 +17,6 @@
 # include "asm.h"
 # include "libftprintf.h"
 
-# include <stdio.h>
-
 /*
 ****************************** INTEGER CONSTANTS *******************************
 */
@@ -30,12 +28,4 @@
 # define EQUAL 0
 # define MALLOC_ERROR -1
 
-/*
-****************************** TYPEDEFS ****************************************
-*/
-
-/*
-****************************** STRUCTURES **************************************
-*/
-
 #endif
diff --git a/asm_dir/model/free.c b/asm_dir/model/free.c
index c8833ce..157c2b2 100644
--- a/asm_dir/model/free.c
+++ b/asm_dir/model/free.c
@@ -18,3 +18,16 @@ int		exit_with_error(t_asm *a, char **line, char *error_message)
 	ft_strdel(line);
 	return (ERROR);
 }
+
+void	free_2d_array(char ***arr)
+{
+	int i;
+
+	i = 0;
+	while (*arr[i])
+	{
+		ft_strdel(&(*arr[i]));
+		i++;
+	}
+	ft_memdel((void **)arr);
+}
diff --git a/vm_dir/.DS_Store b/vm_dir/.DS_Store
index 7220f17..f708a15 100644
Binary files a/vm_dir/.DS_Store and b/vm_dir/.DS_Store differ
diff --git a/vm_dir/music/Rumble.mp3 b/vm_dir/music/Rumble.mp3
deleted file mode 100644
index e39f09c..0000000
Binary files a/vm_dir/music/Rumble.mp3 and /dev/null differ
diff --git a/vm_dir/music/best.mp3 b/vm_dir/music/best.mp3
deleted file mode 100644
index f637f2a..0000000
Binary files a/vm_dir/music/best.mp3 and /dev/null differ
diff --git a/vm_dir/music/smash.mp3 b/vm_dir/music/smash.mp3
deleted file mode 100644
index f7eb036..0000000
Binary files a/vm_dir/music/smash.mp3 and /dev/null differ
diff --git a/vm_dir/music/smash2.mp3 b/vm_dir/music/smash2.mp3
deleted file mode 100644
index 281a4fe..0000000
Binary files a/vm_dir/music/smash2.mp3 and /dev/null differ
diff --git a/vm_dir/music/surprise.mp3 b/vm_dir/music/surprise.mp3
deleted file mode 100644
index ce4253e..0000000
Binary files a/vm_dir/music/surprise.mp3 and /dev/null differ
diff --git a/vm_dir/vm_src/.DS_Store b/vm_dir/vm_src/.DS_Store
index fd930c0..449b5c9 100644
Binary files a/vm_dir/vm_src/.DS_Store and b/vm_dir/vm_src/.DS_Store differ
diff --git a/vm_dir/vm_src/check_pc_lives_n_kill.c b/vm_dir/vm_src/check_pc_lives_n_kill.c
index 7999ae1..de62e54 100644
--- a/vm_dir/vm_src/check_pc_lives_n_kill.c
+++ b/vm_dir/vm_src/check_pc_lives_n_kill.c
@@ -87,6 +87,6 @@ void				zero_all_alives_screams(t_vm *vm)
 		}
 	}
 	if (flag == 1 && vm->visual_flag)
-		system("afplay music/smash2.mp3&");
+		system("afplay vm_dir/vm_scr/music/smash2.mp3");
 	zero_all_lives(vm);
 }
diff --git a/vm_dir/vm_src/dump_end_free_functions.c b/vm_dir/vm_src/dump_end_free_functions.c
index 39d6b03..5185cd5 100644
--- a/vm_dir/vm_src/dump_end_free_functions.c
+++ b/vm_dir/vm_src/dump_end_free_functions.c
@@ -70,7 +70,7 @@ void				end_this_game(t_vm *vm)
 	d = vm->players[last_id].player_number * (-1);
 	s = vm->players[last_id].name;
 	if (vm->visual_flag == 1)
-		system("afplay music/best.mp3&");
+		system("afplay vm_dir/vm_scr/music/best.mp3");
 	if (vm->visual_flag == 1)
 	{
 		print_winner(vm, s);
diff --git a/vm_dir/vm_src/main.c b/vm_dir/vm_src/main.c
index cd03e3e..54391ac 100644
--- a/vm_dir/vm_src/main.c
+++ b/vm_dir/vm_src/main.c
@@ -79,7 +79,7 @@ int					main(int ac, char **av)
 	if (vm->visual_flag)
 	{
 		visualizer(vm);
-		system("afplay music/Rumble.mp3&");
+		system("afplay vm_dir/vm_src/music/Rumble.mp3");
 	}
 	else
 		players_intro(vm);
